# 01 — Data Understanding
## Climate Change: Earth Surface Temperature Analysis

**Người thực hiện:** Lương Minh Kiệt — Project Leader  
**Mục tiêu:** Khảo sát cấu trúc, phạm vi thời gian, kiểu dữ liệu và chất lượng ban đầu của 5 file CSV nhiệt độ bề mặt Trái Đất trước khi xây dựng PostgreSQL pipeline và làm sạch dữ liệu.


## 1. Nguyên tắc thực hiện

Notebook này chỉ **đọc** dữ liệu gốc, không sửa, ghi đè hoặc đổi tên cột. Vì tổng dung lượng dữ liệu lớn, việc kiểm tra toàn bộ file được thực hiện theo từng khối (`chunksize`) để giảm mức sử dụng RAM.

Các công việc chính:

1. Tìm đúng 5 file CSV trong `data/raw/`.
2. Kiểm tra sơ bộ tính toàn vẹn của từng file.
3. Xem kích thước, tên cột, kiểu dữ liệu và các dòng mẫu.
4. Quét toàn bộ dữ liệu để tính số dòng, dữ liệu thiếu, dòng trùng, phạm vi ngày và thống kê mô tả.
5. Ghi lại các vấn đề cần chuyển giao cho notebook PostgreSQL và Data Cleaning.


In [1]:
from pathlib import Path
import csv
import time
import warnings

import numpy as np
import pandas as pd

try:
    from IPython.display import display
except ImportError:
    display = print

pd.set_option('display.max_columns', 30)
pd.set_option('display.max_colwidth', 100)
pd.set_option('display.float_format', lambda value: f'{value:,.4f}')
warnings.filterwarnings('ignore', category=FutureWarning)

EXPECTED_FILES = [
    'GlobalLandTemperaturesByCity.csv',
    'GlobalLandTemperaturesByCountry.csv',
    'GlobalLandTemperaturesByMajorCity.csv',
    'GlobalLandTemperaturesByState.csv',
    'GlobalTemperatures.csv',
]
CHUNK_SIZE = 250_000
print('Pandas version:', pd.__version__)
print('Chunk size:', f'{CHUNK_SIZE:,}', 'rows')


Pandas version: 3.0.3
Chunk size: 250,000 rows


## 2. Xác định thư mục dữ liệu

Trong cấu trúc chuẩn của project, 5 file cần nằm trong `data/raw/`. Hàm bên dưới cũng kiểm tra một số vị trí thường gặp để notebook vẫn chạy khi mở từ thư mục `notebooks/` hoặc khi thử nghiệm trong workspace. Nếu thiếu file, chương trình dừng với thông báo rõ ràng thay vì âm thầm bỏ qua.


In [2]:
def find_data_directory(expected_files):
    candidates = [
        Path('data/raw'),
        Path('../data/raw'),
        Path('../../data/raw'),
        Path('upload'),
        Path('../upload'),
    ]

    checked = []
    for candidate in candidates:
        candidate = candidate.resolve()
        checked.append(str(candidate))
        if candidate.is_dir() and all((candidate / name).is_file() for name in expected_files):
            return candidate

    message = (
        'Không tìm thấy đầy đủ 5 file CSV. Hãy đặt chúng trong data/raw/.\n'
        + 'Các vị trí đã kiểm tra:\n- ' + '\n- '.join(checked)
    )
    raise FileNotFoundError(message)

DATA_DIR = find_data_directory(EXPECTED_FILES)
FILE_PATHS = {name: DATA_DIR / name for name in EXPECTED_FILES}
print('Đang đọc dữ liệu tại:', DATA_DIR)
print('Đã tìm thấy:', len(FILE_PATHS), '/', len(EXPECTED_FILES), 'file')


Đang đọc dữ liệu tại: C:\Users\My Laptop\Desktop\duan1\Global-Surface-Temperature-Analysis\data\raw
Đã tìm thấy: 5 / 5 file


## 3. Kiểm tra sơ bộ tính toàn vẹn của file

CSV hoàn chỉnh thường kết thúc bằng ký tự xuống dòng và dòng cuối phải có cùng số trường với header. Đây là kiểm tra nhanh, hữu ích để phát hiện file tải xuống hoặc tải lên bị gián đoạn. Nó không thay thế checksum từ nguồn dữ liệu chính thức.


In [3]:
def inspect_file_integrity(path, tail_bytes=65_536):
    size_bytes = path.stat().st_size
    with path.open('rb') as file:
        header_bytes = file.readline()
        if size_bytes > tail_bytes:
            file.seek(-tail_bytes, 2)
        tail = file.read()

    header_text = header_bytes.decode('utf-8-sig', errors='replace').rstrip('\r\n')
    header_field_count = len(next(csv.reader([header_text])))
    lines = tail.decode('utf-8', errors='replace').splitlines()
    last_line = lines[-1] if lines else ''
    last_field_count = len(next(csv.reader([last_line]))) if last_line else 0
    ends_with_newline = tail.endswith(b'\n') or tail.endswith(b'\r')
    tail_is_complete = ends_with_newline and last_field_count == header_field_count

    return {
        'file': path.name,
        'size_mb': size_bytes / (1024 ** 2),
        'header_fields': header_field_count,
        'last_row_fields': last_field_count,
        'ends_with_newline': ends_with_newline,
        'tail_status': 'OK' if tail_is_complete else 'CẢNH BÁO: dòng cuối không hoàn chỉnh',
        'last_row_preview': last_line[:120],
    }

integrity_df = pd.DataFrame(inspect_file_integrity(path) for path in FILE_PATHS.values())
display(integrity_df)

problem_files = integrity_df.loc[integrity_df['tail_status'] != 'OK', 'file'].tolist()
if problem_files:
    print('\nCẢNH BÁO: Hãy tải lại các file sau từ nguồn gốc trước bước PostgreSQL/cleaning:')
    for name in problem_files:
        print('-', name)


,file,size_mb,header_fields,last_row_fields,ends_with_newline,tail_status,last_row_preview
0,GlobalLandTemperaturesByCity.csv,297.8377,7,7,True,OK,"2013-09-01,25.793,,Zuwarah,Libya,32.95N,12.45E"
1,GlobalLandTemperaturesByCountry.csv,21.6297,4,4,True,OK,"2013-09-01,,,Zimbabwe"
2,GlobalLandTemperaturesByMajorCity.csv,13.4834,7,7,True,OK,"2013-09-01,,,Xian,China,34.56N,108.97E"
3,GlobalLandTemperaturesByState.csv,29.3447,5,5,True,OK,"2013-09-01,,,Zhejiang,China"
4,GlobalTemperatures.csv,0.1963,9,9,True,OK,"2015-12-01,5.518,0.1,10.725,0.154,0.2869999999999997,0.099,14.774,0.062"


### Phát hiện trên bộ file được cung cấp

`GlobalLandTemperaturesByMajorCity.csv` hiện kết thúc giữa một bản ghi: `1886-10-01,27.427,0.591,Su`. File không có ký tự xuống dòng cuối và bản ghi cuối chỉ có 4/7 trường. Vì vậy file này có dấu hiệu bị cắt trong quá trình tải; cần tải lại bản đầy đủ trước khi nhập PostgreSQL hoặc phân tích chính thức.


## 4. Cấu trúc, kiểu dữ liệu và dữ liệu mẫu

`nrows=5` chỉ đọc vài dòng đầu, giúp kiểm tra nhanh schema mà không phải nạp toàn bộ file lớn vào RAM. Kiểu `object` của cột `dt` ở giai đoạn này có nghĩa là ngày đang được đọc như chuỗi; việc chuyển sang `datetime` sẽ được thực hiện có kiểm soát khi quét dữ liệu.


In [4]:
schema_rows = []
samples = {}

for name, path in FILE_PATHS.items():
    sample = pd.read_csv(path, nrows=5, low_memory=False)
    samples[name] = sample
    for column, dtype in sample.dtypes.items():
        schema_rows.append({'file': name, 'column': column, 'inferred_dtype': str(dtype)})

schema_df = pd.DataFrame(schema_rows)
display(schema_df)

for name, sample in samples.items():
    print('\n' + '=' * 90)
    print(name, '| sample shape:', sample.shape)
    display(sample)


,file,column,inferred_dtype
0,GlobalLandTemperaturesByCity.csv,dt,str
1,GlobalLandTemperaturesByCity.csv,averagetemperature,float64
2,GlobalLandTemperaturesByCity.csv,averagetemperatureuncertainty,float64
3,GlobalLandTemperaturesByCity.csv,city,str
4,GlobalLandTemperaturesByCity.csv,country,str
5,GlobalLandTemperaturesByCity.csv,latitude,str
6,GlobalLandTemperaturesByCity.csv,longitude,str
7,GlobalLandTemperaturesByCountry.csv,dt,str
8,GlobalLandTemperaturesByCountry.csv,AverageTemperature,float64
9,GlobalLandTemperaturesByCountry.csv,AverageTemperatureUncertainty,float64



GlobalLandTemperaturesByCity.csv | sample shape: (5, 7)


,dt,averagetemperature,averagetemperatureuncertainty,city,country,latitude,longitude
0,1863-01-01,3.2650,2.1900,Århus,Denmark,57.05N,10.33E
1,1863-02-01,3.3180,3.0310,Århus,Denmark,57.05N,10.33E
2,1863-03-01,2.6990,0.6470,Århus,Denmark,57.05N,10.33E
3,1863-04-01,6.7130,1.6630,Århus,Denmark,57.05N,10.33E
4,1863-05-01,10.0280,1.2970,Århus,Denmark,57.05N,10.33E



GlobalLandTemperaturesByCountry.csv | sample shape: (5, 4)


,dt,AverageTemperature,AverageTemperatureUncertainty,Country
0,1743-11-01,4.3840,2.2940,Åland
1,1743-12-01,NaN,NaN,Åland
2,1744-01-01,NaN,NaN,Åland
3,1744-02-01,NaN,NaN,Åland
4,1744-03-01,NaN,NaN,Åland



GlobalLandTemperaturesByMajorCity.csv | sample shape: (5, 7)


,dt,AverageTemperature,AverageTemperatureUncertainty,City,Country,Latitude,Longitude
0,1849-01-01,26.7040,1.4350,Abidjan,Côte D'Ivoire,5.63N,3.23W
1,1849-02-01,27.4340,1.3620,Abidjan,Côte D'Ivoire,5.63N,3.23W
2,1849-03-01,28.1010,1.6120,Abidjan,Côte D'Ivoire,5.63N,3.23W
3,1849-04-01,26.1400,1.3870,Abidjan,Côte D'Ivoire,5.63N,3.23W
4,1849-05-01,25.4270,1.2000,Abidjan,Côte D'Ivoire,5.63N,3.23W



GlobalLandTemperaturesByState.csv | sample shape: (5, 5)


,dt,AverageTemperature,AverageTemperatureUncertainty,State,Country
0,1855-05-01,25.5440,1.1710,Acre,Brazil
1,1855-06-01,24.2280,1.1030,Acre,Brazil
2,1855-07-01,24.3710,1.0440,Acre,Brazil
3,1855-08-01,25.4270,1.0730,Acre,Brazil
4,1855-09-01,25.6750,1.0140,Acre,Brazil



GlobalTemperatures.csv | sample shape: (5, 9)


,dt,LandAverageTemperature,LandAverageTemperatureUncertainty,LandMaxTemperature,LandMaxTemperatureUncertainty,LandMinTemperature,LandMinTemperatureUncertainty,LandAndOceanAverageTemperature,LandAndOceanAverageTemperatureUncertainty
0,1750-01-01,3.0340,3.5740,NaN,NaN,NaN,NaN,NaN,NaN
1,1750-02-01,3.0830,3.7020,NaN,NaN,NaN,NaN,NaN,NaN
2,1750-03-01,5.6260,3.0760,NaN,NaN,NaN,NaN,NaN,NaN
3,1750-04-01,8.4900,2.4510,NaN,NaN,NaN,NaN,NaN,NaN
4,1750-05-01,11.5730,2.0720,NaN,NaN,NaN,NaN,NaN,NaN


## 5. Ý nghĩa các cột quan trọng

| Cột | Ý nghĩa | Lưu ý ban đầu |
|---|---|---|
| `dt` | Thời điểm ghi nhận nhiệt độ, theo tháng | Cần chuyển từ chuỗi sang `datetime`; có thể tách năm/tháng ở Feature Engineering |
| `AverageTemperature` / `averagetemperature` | Nhiệt độ trung bình ước tính, đơn vị °C | Hai cách viết hoa/thường đang tồn tại giữa các file; không tự đổi tên trong dữ liệu raw |
| `AverageTemperatureUncertainty` / `averagetemperatureuncertainty` | Mức độ không chắc chắn của ước tính nhiệt độ, đơn vị °C | Giá trị càng lớn thì độ tin cậy của ước tính càng thấp |
| `City` / `city` | Thành phố | Dùng để nhóm và so sánh xu hướng theo thành phố |
| `State` | Bang/tỉnh/khu vực hành chính | Chỉ xuất hiện trong file theo bang |
| `Country` / `country` | Quốc gia | Cần rà soát cách đặt tên quốc gia trước khi join |
| `Latitude` / `latitude` | Vĩ độ dạng chuỗi có hậu tố `N` hoặc `S` | Cần tách hướng và chuyển thành số có dấu trong Data Cleaning/Feature Engineering |
| `Longitude` / `longitude` | Kinh độ dạng chuỗi có hậu tố `E` hoặc `W` | Cần tách hướng và chuyển thành số có dấu trong Data Cleaning/Feature Engineering |
| `LandAverageTemperature` | Nhiệt độ trung bình toàn cầu trên đất liền | Chỉ có trong `GlobalTemperatures.csv` |
| `LandMaxTemperature`, `LandMinTemperature` | Nhiệt độ cực đại/cực tiểu trung bình trên đất liền | Có giai đoạn lịch sử bị thiếu |
| `LandAndOceanAverageTemperature` | Nhiệt độ trung bình kết hợp đất liền và đại dương | Có phạm vi thời gian ngắn hơn nhiệt độ đất liền |

Các cột có hậu tố `Uncertainty` luôn biểu diễn độ không chắc chắn tương ứng với đại lượng nhiệt độ đứng trước nó.


## 6. Quét toàn bộ dữ liệu theo từng khối

Hàm dưới đây thực hiện một lượt quét mỗi file và tạo ra:

- Số dòng, số cột và phạm vi thời gian.
- Số lượng và tỷ lệ phần trăm dữ liệu thiếu.
- Số dòng trùng dựa trên hash 64-bit của toàn bộ bản ghi. Xác suất va chạm hash là rất nhỏ.
- Số giá trị phân biệt của City/Country/State.
- Thống kê chính xác `count`, `mean`, `std`, `min`, `max` cho cột số bằng thuật toán gộp từng khối.

Không tính tứ phân vị chính xác trong lượt quét này vì việc đó thường yêu cầu giữ toàn bộ cột số trong bộ nhớ. Các phân vị và biểu đồ phân phối phù hợp hơn với notebook `04_eda_visualization.ipynb`.


In [7]:
def merge_numeric_state(state, values):
    values = values.dropna().astype('float64')
    if values.empty:
        return state

    n_b = int(values.count())
    mean_b = float(values.mean())
    m2_b = float(((values - mean_b) ** 2).sum())
    min_b = float(values.min())
    max_b = float(values.max())

    if state is None:
        return {'count': n_b, 'mean': mean_b, 'm2': m2_b, 'min': min_b, 'max': max_b}

    n_a = state['count']
    delta = mean_b - state['mean']
    n_total = n_a + n_b
    state['mean'] += delta * n_b / n_total
    state['m2'] += m2_b + delta ** 2 * n_a * n_b / n_total
    state['count'] = n_total
    state['min'] = min(state['min'], min_b)
    state['max'] = max(state['max'], max_b)
    return state


def scan_csv(path, chunk_size=250_000):
    sample = pd.read_csv(path, nrows=10_000, low_memory=False)
    columns = sample.columns.tolist()
    numeric_columns = sample.select_dtypes(include=np.number).columns.tolist()
    dimension_columns = [c for c in columns if c.casefold() in {'city', 'country', 'state'}]

    row_count = 0
    missing_counts = pd.Series(0, index=columns, dtype='int64')
    invalid_date_count = 0
    date_min = None
    date_max = None
    numeric_states = {column: None for column in numeric_columns}
    unique_values = {column: set() for column in dimension_columns}
    row_hashes = []

    for chunk in pd.read_csv(path, chunksize=chunk_size, low_memory=False):
        row_count += len(chunk)
        missing_counts = missing_counts.add(chunk.isna().sum(), fill_value=0).astype('int64')
        row_hashes.append(pd.util.hash_pandas_object(chunk, index=False).to_numpy(dtype='uint64'))

        if 'dt' in chunk.columns:
            parsed_dates = pd.to_datetime(chunk['dt'], errors='coerce')
            invalid_date_count += int(parsed_dates.isna().sum() - chunk['dt'].isna().sum())
            current_min = parsed_dates.min()
            current_max = parsed_dates.max()
            if pd.notna(current_min) and (date_min is None or current_min < date_min):
                date_min = current_min
            if pd.notna(current_max) and (date_max is None or current_max > date_max):
                date_max = current_max

        for column in numeric_columns:
            numeric_states[column] = merge_numeric_state(numeric_states[column], chunk[column])

        for column in dimension_columns:
            unique_values[column].update(chunk[column].dropna().astype(str).unique())

    all_hashes = np.concatenate(row_hashes) if row_hashes else np.array([], dtype='uint64')
    duplicate_rows = int(len(all_hashes) - len(np.unique(all_hashes)))

    numeric_summary = []
    for column, state in numeric_states.items():
        if state is None:
            continue
        std = np.sqrt(state['m2'] / (state['count'] - 1)) if state['count'] > 1 else 0.0
        numeric_summary.append({
            'column': column, 'count': state['count'], 'mean': state['mean'],
            'std': std, 'min': state['min'], 'max': state['max'],
        })

    return {
        'file': path.name,
        'rows': row_count,
        'columns': len(columns),
        'column_names': columns,
        'dtypes': sample.dtypes.astype(str).to_dict(),
        'date_min': date_min,
        'date_max': date_max,
        'invalid_dates': invalid_date_count,
        'missing_counts': missing_counts.to_dict(),
        'duplicate_rows': duplicate_rows,
        'unique_counts': {column: len(values) for column, values in unique_values.items()},
        'numeric_summary': numeric_summary,
    }


In [8]:
scan_results = []
start_time = time.perf_counter()

for name, path in FILE_PATHS.items():
    file_start = time.perf_counter()
    result = scan_csv(path, chunk_size=CHUNK_SIZE)
    scan_results.append(result)
    elapsed = time.perf_counter() - file_start
    print(f'Đã quét {name}: {result["rows"]:,} dòng trong {elapsed:,.2f} giây')

print(f'\nTổng thời gian quét: {time.perf_counter() - start_time:,.2f} giây')


Đã quét GlobalLandTemperaturesByCity.csv: 5,010,113 dòng trong 9.49 giây
Đã quét GlobalLandTemperaturesByCountry.csv: 577,462 dòng trong 0.74 giây
Đã quét GlobalLandTemperaturesByMajorCity.csv: 239,177 dòng trong 0.37 giây
Đã quét GlobalLandTemperaturesByState.csv: 645,675 dòng trong 0.95 giây
Đã quét GlobalTemperatures.csv: 3,192 dòng trong 0.02 giây

Tổng thời gian quét: 11.57 giây


## 7. Tổng quan 5 bộ dữ liệu

Bảng này giúp so sánh quy mô và phạm vi thời gian. File theo thành phố là bộ lớn nhất; `GlobalTemperatures.csv` nhỏ nhất nhưng có chuỗi thời gian toàn cầu đến năm 2015.


In [9]:
overview_rows = []
for result in scan_results:
    overview_rows.append({
        'file': result['file'],
        'rows': result['rows'],
        'columns': result['columns'],
        'date_min': result['date_min'].date() if result['date_min'] is not None else None,
        'date_max': result['date_max'].date() if result['date_max'] is not None else None,
        'invalid_dates': result['invalid_dates'],
        'duplicate_rows': result['duplicate_rows'],
    })

overview_df = pd.DataFrame(overview_rows)
display(overview_df)
print('Tổng số bản ghi của 5 file:', f'{overview_df["rows"].sum():,}')


,file,rows,columns,date_min,date_max,invalid_dates,duplicate_rows
0,GlobalLandTemperaturesByCity.csv,5010113,7,1863-01-01,2013-09-01,0,0
1,GlobalLandTemperaturesByCountry.csv,577462,4,1743-11-01,2013-09-01,0,0
2,GlobalLandTemperaturesByMajorCity.csv,239177,7,1743-11-01,2013-09-01,0,0
3,GlobalLandTemperaturesByState.csv,645675,5,1743-11-01,2013-09-01,0,0
4,GlobalTemperatures.csv,3192,9,1750-01-01,2015-12-01,0,0


Tổng số bản ghi của 5 file: 6,475,619


## 8. Kiểm tra dữ liệu thiếu

Không xử lý giá trị thiếu ở notebook này. Mục đích là đo lường và cung cấp bằng chứng cho việc xây dựng quy tắc làm sạch ở `03_data_cleaning.ipynb`. Tỷ lệ thiếu cần được đọc cùng với phạm vi thời gian: một số biến toàn cầu chỉ bắt đầu được ghi nhận muộn hơn nên thiếu có tính hệ thống, không hẳn là lỗi ngẫu nhiên.


In [10]:
missing_rows = []
for result in scan_results:
    for column, count in result['missing_counts'].items():
        missing_rows.append({
            'file': result['file'],
            'column': column,
            'missing_count': int(count),
            'missing_percent': count / result['rows'] * 100 if result['rows'] else 0,
        })

missing_df = pd.DataFrame(missing_rows)
missing_only_df = missing_df.loc[missing_df['missing_count'] > 0].sort_values(
    ['file', 'missing_percent'], ascending=[True, False]
)
display(missing_only_df)

if missing_only_df.empty:
    print('Không phát hiện giá trị thiếu.')
else:
    print('Các cột không xuất hiện trong bảng trên có 0 giá trị thiếu.')


,file,column,missing_count,missing_percent
2,GlobalLandTemperaturesByCity.csv,averagetemperatureuncertainty,47369,0.9455
8,GlobalLandTemperaturesByCountry.csv,AverageTemperature,32651,5.6542
9,GlobalLandTemperaturesByCountry.csv,AverageTemperatureUncertainty,31912,5.5263
12,GlobalLandTemperaturesByMajorCity.csv,AverageTemperature,11002,4.5999
13,GlobalLandTemperaturesByMajorCity.csv,AverageTemperatureUncertainty,11002,4.5999
19,GlobalLandTemperaturesByState.csv,AverageTemperature,25648,3.9723
20,GlobalLandTemperaturesByState.csv,AverageTemperatureUncertainty,25648,3.9723
26,GlobalTemperatures.csv,LandMaxTemperature,1200,37.5940
27,GlobalTemperatures.csv,LandMaxTemperatureUncertainty,1200,37.5940
28,GlobalTemperatures.csv,LandMinTemperature,1200,37.5940


Các cột không xuất hiện trong bảng trên có 0 giá trị thiếu.


## 9. Kiểm tra dòng trùng

Dòng trùng được xét trên toàn bộ giá trị của một bản ghi. Kết quả 0 không có nghĩa là mỗi khóa logic `(dt, địa điểm)` đều duy nhất; kiểm tra khóa nghiệp vụ nên được thực hiện thêm khi thiết kế bảng PostgreSQL.


In [11]:
duplicate_df = overview_df[['file', 'rows', 'duplicate_rows']].copy()
duplicate_df['duplicate_percent'] = duplicate_df['duplicate_rows'] / duplicate_df['rows'] * 100
display(duplicate_df)


,file,rows,duplicate_rows,duplicate_percent
0,GlobalLandTemperaturesByCity.csv,5010113,0,0.0000
1,GlobalLandTemperaturesByCountry.csv,577462,0,0.0000
2,GlobalLandTemperaturesByMajorCity.csv,239177,0,0.0000
3,GlobalLandTemperaturesByState.csv,645675,0,0.0000
4,GlobalTemperatures.csv,3192,0,0.0000


## 10. Thống kê mô tả các cột số

`mean` là giá trị trung bình, `std` là độ lệch chuẩn, còn `min` và `max` cho biết biên quan sát. Giá trị cực trị ở đây chỉ là dấu hiệu cần khảo sát; chưa được kết luận là outlier và chưa bị loại bỏ.


In [12]:
numeric_rows = []
for result in scan_results:
    for row in result['numeric_summary']:
        numeric_rows.append({'file': result['file'], **row})

numeric_summary_df = pd.DataFrame(numeric_rows)
display(numeric_summary_df)


,file,column,count,mean,std,min,max
0,GlobalLandTemperaturesByCity.csv,averagetemperature,5010113,17.3489,10.2452,-42.7040,38.1820
1,GlobalLandTemperaturesByCity.csv,averagetemperatureuncertainty,4962744,0.5868,0.4528,0.0340,12.6940
2,GlobalLandTemperaturesByCountry.csv,AverageTemperature,544811,17.1934,10.9540,-37.6580,38.8420
3,GlobalLandTemperaturesByCountry.csv,AverageTemperatureUncertainty,545550,1.0191,1.2019,0.0520,15.0030
4,GlobalLandTemperaturesByMajorCity.csv,AverageTemperature,228175,18.1260,10.0248,-26.7720,38.2830
5,GlobalLandTemperaturesByMajorCity.csv,AverageTemperatureUncertainty,228175,0.9693,0.9796,0.0400,14.0370
6,GlobalLandTemperaturesByState.csv,AverageTemperature,620027,8.9931,13.7722,-45.3890,36.3390
7,GlobalLandTemperaturesByState.csv,AverageTemperatureUncertainty,620027,1.2876,1.3604,0.0360,12.6460
8,GlobalTemperatures.csv,LandAverageTemperature,3180,8.3747,4.3813,-2.0800,19.0210
9,GlobalTemperatures.csv,LandAverageTemperatureUncertainty,3180,0.9385,1.0964,0.0340,7.8800


## 11. Số lượng địa điểm phân biệt

Số lượng giá trị phân biệt giúp kiểm tra độ phủ địa lý và dự kiến kích thước các bảng dimension trong PostgreSQL. File bị cắt có thể làm sai kết quả đếm, vì vậy số liệu của Major City chỉ mang tính tạm thời cho đến khi thay file đầy đủ.


In [13]:
dimension_rows = []
for result in scan_results:
    for column, count in result['unique_counts'].items():
        dimension_rows.append({'file': result['file'], 'dimension': column, 'unique_count': count})

dimension_df = pd.DataFrame(dimension_rows)
display(dimension_df)


,file,dimension,unique_count
0,GlobalLandTemperaturesByCity.csv,city,2729
1,GlobalLandTemperaturesByCity.csv,country,80
2,GlobalLandTemperaturesByCountry.csv,Country,243
3,GlobalLandTemperaturesByMajorCity.csv,City,100
4,GlobalLandTemperaturesByMajorCity.csv,Country,49
5,GlobalLandTemperaturesByState.csv,State,241
6,GlobalLandTemperaturesByState.csv,Country,7


## 12. So sánh schema giữa các file

Bộ file hiện tại không thống nhất cách viết hoa tên cột. Đặc biệt, file theo City dùng `averagetemperature`, `city`, `country`, `latitude`, `longitude`, trong khi các file tương tự dùng PascalCase. Việc chuẩn hóa phải được ghi rõ trong notebook Cleaning hoặc trong SQL, không sửa trực tiếp file raw.


In [14]:
all_columns = sorted({column for result in scan_results for column in result['column_names']}, key=str.casefold)
schema_matrix = pd.DataFrame(index=all_columns)
for result in scan_results:
    short_name = result['file'].replace('.csv', '')
    present = set(result['column_names'])
    schema_matrix[short_name] = ['✓' if column in present else '' for column in all_columns]

display(schema_matrix)


,GlobalLandTemperaturesByCity,GlobalLandTemperaturesByCountry,GlobalLandTemperaturesByMajorCity,GlobalLandTemperaturesByState,GlobalTemperatures
AverageTemperature,,✓,✓,✓,
averagetemperature,✓,,,,
averagetemperatureuncertainty,✓,,,,
AverageTemperatureUncertainty,,✓,✓,✓,
City,,,✓,,
city,✓,,,,
Country,,✓,✓,✓,
country,✓,,,,
dt,✓,✓,✓,✓,✓
LandAndOceanAverageTemperature,,,,,✓


## 13. Nhận xét ban đầu từ 5 file đã nhận

### 13.1 Quy mô và thời gian

- `GlobalLandTemperaturesByCity.csv`: **5.010.113 dòng**, 7 cột, từ **1863-01** đến **2013-09**; gồm 2.729 giá trị city và 80 country.
- `GlobalLandTemperaturesByCountry.csv`: **577.462 dòng**, 4 cột, từ **1743-11** đến **2013-09**; gồm 243 country.
- `GlobalLandTemperaturesByMajorCity.csv`: pandas đọc được **212.429 dòng**, 7 cột, nhưng file bị cắt giữa bản ghi cuối nên số liệu chưa đáng tin cậy cho phân tích chính thức.
- `GlobalLandTemperaturesByState.csv`: **645.675 dòng**, 5 cột, từ **1743-11** đến **2013-09**; gồm 241 state thuộc 7 country.
- `GlobalTemperatures.csv`: **3.192 dòng**, 9 cột, từ **1750-01** đến **2015-12**.
- Tổng cộng pandas đọc được **6.448.871 bản ghi** từ 5 file hiện tại.

### 13.2 Chất lượng dữ liệu

- Không phát hiện dòng trùng hoàn toàn trong 5 file bằng row hash.
- File City thiếu 47.369 giá trị `averagetemperatureuncertainty` (khoảng 0,95%), nhưng không thiếu `averagetemperature`. Điều này cùng với schema viết thường và mốc bắt đầu 1863 cho thấy cần xác minh đây có phải bản raw gốc hay một bản đã được tiền xử lý.
- File Country thiếu 32.651 `AverageTemperature` (5,65%) và 31.912 `AverageTemperatureUncertainty` (5,53%).
- File State thiếu đồng thời 25.648 nhiệt độ và uncertainty (3,97%).
- File Global thiếu 1.200 giá trị (37,59%) ở các biến max/min và land-ocean. Đây có khả năng là thiếu theo giai đoạn lịch sử vì các phép đo này bắt đầu muộn hơn, nên không nên điền trung bình một cách máy móc.
- Major City có 1 dòng cuối thiếu `Country`, `Latitude`, `Longitude` do bản ghi bị cắt, không phải dữ liệu thiếu tự nhiên.

### 13.3 Giá trị nhiệt độ và uncertainty

- Nhiệt độ trung bình theo City nằm trong khoảng **-42,704°C đến 38,182°C**; theo Country từ **-37,658°C đến 38,842°C**; theo State từ **-45,389°C đến 36,339°C**.
- Uncertainty có thể cao đến khoảng **15°C** ở file Country, thường liên quan đến các quan sát lịch sử xa hơn. Nên phân tích uncertainty theo năm trước khi quyết định lọc hay gán trọng số.
- Các giá trị cực trị chưa được xem là sai. Chúng cần được kiểm tra theo địa điểm, thời gian và uncertainty trong notebook EDA/Cleaning.


## 14. Kết luận và bàn giao cho các bước tiếp theo

1. **Ưu tiên tải lại `GlobalLandTemperaturesByMajorCity.csv`** vì file hiện tại bị cắt giữa dòng. Không nhập file này vào PostgreSQL cho đến khi kiểm tra lại.
2. Xác minh nguồn của file City vì schema viết thường, không thiếu nhiệt độ và chỉ bắt đầu từ năm 1863; các đặc điểm này khác các file còn lại.
3. Ở `02_postgresql_pipeline.ipynb`, cần đọc cấu hình từ biến môi trường, tạo bảng theo đúng schema thật và kiểm tra row count sau import.
4. Ở `03_data_cleaning.ipynb`, cần chuẩn hóa tên cột có ghi log, chuyển `dt` sang datetime, xử lý latitude/longitude có hướng, xác định quy tắc missing theo từng file và ghi số dòng trước/sau xử lý.
5. Không sửa file trong `data/raw/`; mọi dữ liệu đã làm sạch phải được tạo lại bằng code và lưu trong `data/processed/`.

**Kết quả:** Notebook đã hoàn thành bước Data Understanding cho toàn bộ 5 file được cung cấp, đồng thời phát hiện một lỗi toàn vẹn nghiêm trọng cần xử lý trước pipeline tiếp theo.
